// mongodb-rag-ingest/build/sources/RstOnGitHubDataSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeRstOnGitHubDataSource = void 0;
const rstToSnootyAst_1 = require("./snooty/rstToSnootyAst");
const snootyAstToMd_1 = require("./snooty/snootyAstToMd");
const GitHubDataSource_1 = require("./GitHubDataSource");
const extractMarkdownH1_1 = require("./extractMarkdownH1");
/**
  Loads an rST docs site from a GitHub repo.
 */
const makeRstOnGitHubDataSource = ({ name, repoUrl, repoLoaderOptions, pathToPageUrl, getMetadata, }) => {
    return (0, GitHubDataSource_1.makeGitHubDataSource)({
        name,
        repoUrl,
        repoLoaderOptions: {
            ...(repoLoaderOptions ?? {}),
            ignoreFiles: [
                /^(?!.*\.rst$).*/i, // Anything BUT .rst extension
                ...(repoLoaderOptions?.ignoreFiles ?? []),
            ],
        },
        async handleDocumentInRepo(document) {
            const { source } = document.metadata;
            const url = pathToPageUrl(source);
            const body = (0, snootyAstToMd_1.snootyAstToMd)((0, rstToSnootyAst_1.rstToSnootyAst)(document.pageContent));
            const title = (0, extractMarkdownH1_1.extractMarkdownH1)(body);
            const metadata = getMetadata
                ? getMetadata({ document, url, bodyMarkdown: body, title })
                : undefined;
            return {
                body,
                format: "md",
                sourceName: name,
                url,
                metadata,
                title,
            };
        },
    });
};
exports.makeRstOnGitHubDataSource = makeRstOnGitHubDataSource;
