// mongodb-rag-ingest/build/commands/all.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.doAllCommand = void 0;
const mongodb_rag_core_1 = require("mongodb-rag-core");
const pages_1 = require("./pages");
const embed_1 = require("./embed");
const withConfig_1 = require("../withConfig");
const commandModule = {
    command: "all",
    builder(args) {
        return (0, withConfig_1.withConfigOptions)(args);
    },
    async handler(args) {
        return (0, withConfig_1.withConfig)(exports.doAllCommand, {
            ...args,
            doPagesCommand: pages_1.doPagesCommand,
        });
    },
    describe: "Run 'pages' and 'embed' since last successful run",
};
exports.default = commandModule;
const doAllCommand = async (config, { doPagesCommand, }) => {
    const { ingestMetaStore } = config;
    const lastSuccessfulRunDate = await ingestMetaStore.loadLastSuccessfulRunDate();
    mongodb_rag_core_1.logger.info(`Last successful run date: ${lastSuccessfulRunDate}`);
    await doPagesCommand(config, {});
    await (0, embed_1.doEmbedCommand)(config, {
        since: lastSuccessfulRunDate ?? new Date("2023-01-01"),
    });
    mongodb_rag_core_1.logger.info(`Updating last successful run date`);
    await ingestMetaStore.updateLastSuccessfulRunDate();
};
exports.doAllCommand = doAllCommand;
