// mongodb-rag-ingest/build/commands/pages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.doPagesCommand = void 0;
const mongodb_rag_core_1 = require("mongodb-rag-core");
const updatePages_1 = require("../pages/updatePages");
const withConfig_1 = require("../withConfig");
const commandModule = {
    command: "pages",
    builder(args) {
        return (0, withConfig_1.withConfigOptions)(args).option("source", {
            string: true,
            description: "A source name to load. If unspecified, loads all sources.",
        });
    },
    handler(args) {
        return (0, withConfig_1.withConfig)(exports.doPagesCommand, args);
    },
    describe: "Update pages data from sources",
};
exports.default = commandModule;
const doPagesCommand = async ({ pageStore, dataSources }, { source }) => {
    const requestedSources = new Set(Array.isArray(source) ? source : [source]);
    const sources = source === undefined
        ? dataSources
        : dataSources.filter(({ name }) => requestedSources.has(name));
    if (sources.length === 0) {
        throw new Error(`Request at least one valid source. Available sources:\n${dataSources
            .map(({ name }) => `- ${name}`)
            .join("\n")}`);
    }
    mongodb_rag_core_1.logger.info(`Loaded sources:\n${sources.map(({ name }) => `- ${name}`).join("\n")}`);
    await (0, updatePages_1.updatePages)({
        sources,
        pageStore,
    });
};
exports.doPagesCommand = doPagesCommand;
